### ISSUE TITLE
Importing a friend function prototype is a mess

### ISSUE THREAD

**Issue by martong** [Ericsson/clang#327]
Importing the prototype B::f here:
```
struct B { friend void f(); };
void f() {}
```

Results a messy AST, doubled definition, definition inside the class. It is so wrong :(
```
`-CXXRecordDecl 0x24d72f8 <input.cc:2:9, col:16> col:16 struct B definition
  |-FunctionDecl 0x24d7500 parent 0x24a1210 <col:20, line:3:19> line:2:32 f 'void (void)'
  | `-CompoundStmt 0x24d75a8 <line:3:18, col:19>
  |-CXXRecordDecl 0x24d7418 <line:2:9, col:16> col:16 implicit struct B
  `-FriendDecl 0x24d75c0 <col:20, line:3:19> line:2:32
    `-FunctionDecl 0x24d7500 parent 0x24a1210 <col:20, line:3:19> line:2:32 f 'void (void)'
      `-CompoundStmt 0x24d75a8 <line:3:18, col:19>
```

**Comment by balazske** [comment]
The same FunctionDecl object is inserted in the CXXRecordDecl and into the FriendDecl. Normally it should be only in the FriendDecl. It looks like that the definition can appear inside the friend function (like `friend void f(){}`) so this is probably not a problem. (But better is to not move the definition from one place to other, specially not into a friend prototype.)

**Comment by baloghadamsoftware** [comment]
Is the bug related to this assertion?

clang-5.0: /home/edmbalo/ctu/llvm/tools/clang/include/clang/AST/DeclBase.h:1041: void clang::Decl::setObjectOfFriendDecl(bool): Assertion `!(OldNS & ~(IDNS_Tag | IDNS_Ordinary | IDNS_Type | IDNS_TagFriend | IDNS_OrdinaryFriend | IDNS_LocalExtern)) && "namespace includes other than ordinary or tag"' failed.

**Comment by martong** [comment]
Here is the test case:
```
TEST_F(ImportFunctions, ImportPrototypeOfFriendShouldBringInTheRedeclChain) {
  auto Code =
      R"(
        struct B { friend void f(); };
        void f() {}
        )";
  auto Pattern = functionDecl(hasName("f"));
  Decl *FromTU = getTuDecl(Code, Lang_CXX);
  auto *Proto = FirstDeclMatcher<FunctionDecl>().match(FromTU, Pattern);
  ASSERT_TRUE(!Proto->doesThisDeclarationHaveABody());
  auto *Def = LastDeclMatcher<FunctionDecl>().match(FromTU, Pattern);
  ASSERT_TRUE(Def->doesThisDeclarationHaveABody());

  auto *ToProto = cast<FunctionDecl>(Import(Proto, Lang_CXX));
  EXPECT_EQ(DeclCounter<FunctionDecl>().match(ToProto->getTranslationUnitDecl(),
                                              Pattern),
            2u);
  EXPECT_TRUE(!ToProto->doesThisDeclarationHaveABody());
  EXPECT_TRUE(ToProto->hasBody());
  const FunctionDecl* ToDef = nullptr;
  ToProto->getBody(ToDef);
  EXPECT_EQ(ToDef->getPreviousDecl(), ToProto);
}
```

**Comment by martong** [comment]
@baloghadamsoftware This is not related. I think the solution to that assertion is to remove the call of `setObjectOfFriendDecl` because the preceding `Import(FriendD)` already set up the `IdentifierNamespace`.

**Comment by Xazax-hun** [comment]
Maybe there is a problem with the current lookup logic not checking friend namespace?

**Comment by martong** [comment]
Perhaps. But I think it is more complex than that. One very important issue I just found about friends is that an in-class friend definition is not available to normal lookup. 
E.g.
```
struct X {
  friend void f() {};
};
void g() { f(); } // normal lookup can't find f according to C++ lookup rules.
```

In order to make the lookup to find f, one has to declare that in the enclosing scope:
```
struct X {
  friend void f() {};
};
void f();
```

Consequently, the structure of a redecl chain is indeed really important, at least in this case.
The current import logic would not import the second prototype which actually makes `f` available in g().

**Comment by martong** [comment]
Fixed in #329
